### PR TITLE
[Technical Steering] Update Contribution Process

### LINKED ISSUE #630: [Technical Steering] Update Contribution process
There is a need to update contribution process:
- Introducing new labels.
- Policy around Idle PRs.
  - PRs that are not moving due to:
    - Lack of engagement from the person who created the PR.
    - Lack of reviewers engagement.

### PR DESCRIPTION
**Pull request by rabi-abdel**
Fixes #630

### CODE REVIEW

**Comment by rabi-abdel** [comment]
> We should have Unresolved label to cater for the situation requiring breaking the deadlock

done.

**Comment by rabi-abdel** [comment]
> @rabi-abdel: Shouldn't it be "contributor" instead of "committer" in almost all occurences?
> 
> That's how github attributes the people and how we are listed in https://wiki.lfnetworking.org/display/LN/CNTT+Workstream+Signup
> 
> As "committers" I would only count people you have done a commit in a PR.

@hdamker change was made as suggested

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,65 +1,99 @@
...
+  - **Backlog**: This label applies to issues only to indicate that the issue is to be added to a backlog of features for future releases (to be considered during technical steering meeting).
+  - **Workstream Name**: This label applies to both issues and PRs to indicate which project the issue/PR is targeting.
+  - **Chapter Number**: This label applies to PRs (and optionally to issues) to indicate which chapter the PR is targeting.
+  - **General**: This label applies to both issues and PRs to indicate that this issue/PR is not corresponding to any particular chapter/file.
+  - **Major/Minor**: This label applies to both issues and PRs to indicate if the issue/PR is addressing a major or minor change. 
+  - **WIP**: This label applies to PRs to indicate that this PR is still being developed and no reviews/approvals are still expected.
+  - **Enhancement**: This label applies to both issues and PRs to indicate that the issue/Pr is proposing an Enhancement.
+  - **Returned**: This label applies to both issues and PRs to indicate that the issue/PR has been returned for reasons such as (more information needed, not a high priority, no longer valid, etc).
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: What "returned" means is not defined (circular definition)

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,65 +1,99 @@
...
+
+  - **Roadmap Item**: This label applies to both issues and PRs to indicate that the issue/PR is for a roadmap item for an upcoming release.
+  - **Fix needed**: This label applies to both issues and PRs to indicate that the issue/PR is to fix an existing bug.
+  - **Backlog**: This label applies to issues only to indicate that the issue is to be added to a backlog of features for future releases (to be considered during technical steering meeting).
+  - **Workstream Name**: This label applies to both issues and PRs to indicate which project the issue/PR is targeting.
+  - **Chapter Number**: This label applies to PRs (and optionally to issues) to indicate which chapter the PR is targeting.
+  - **General**: This label applies to both issues and PRs to indicate that this issue/PR is not corresponding to any particular chapter/file.
+  - **Major/Minor**: This label applies to both issues and PRs to indicate if the issue/PR is addressing a major or minor change. 
+  - **WIP**: This label applies to PRs to indicate that this PR is still being developed and no reviews/approvals are still expected.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: typo: please delete "still"

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,65 +1,99 @@
...
 
+- All issues must be clearly titled, described, and tagged with the right labels when created.
+  - Title needs to be in the following format [WSx Chx].
+    - WSx: The name of the work stream it is targeting.
+    - Chx: The name of the chapter it is targeting (if applicable).
+- Issues should be correctly set with the correct Project and Milestone.
+  - Milestone = the upcoming release related to the issue/PR.
+  - Project = Work stream.
+- Workstream Leads has the sole responsibility to assign issues to individuals to address them.
+- **self assignment** are NOT recommended. Contributors may ask WSL (vai comment or others) to get the issue assigned to them if they would like to work on it.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: typo: via

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,65 +1,99 @@
...
+  - This is to make sure items are worked on based on priorities and right reasoning.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: tpyo: extra "on" in "...  are worked on based on .." ---> "... are worked based on  ..."

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,65 +1,99 @@
...
+- WSL may label an issue as "**Returned**" to request further information or to decide not to work on it.
 
 <a name="prs"></a>
 ### Pull Requests
-- PRs must only be created by a committer who has an issue assigned to him/her by the respective WSL.
+- PRs must only be created by a contributor who has an issue assigned to them by the respective WSL.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: Should this be more specific, like in, "A PR for an issues must only be created by the contributor assigned to the issue by the respective WSL"

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,65 +1,99 @@
...
+  - 2 business days cool off period should be applied before Final approval.
   - Final Approval by TSL (to make sure process is followed)
-- Only One approval is permitted per each organization for a given PR.
-- The selection of which Committers to approve a PR is made by the WSL and should take those factors into consideration:
-  - Committers needs to be actively discussing the PR to be selected for approval.
-  - Committers needs to be active in the workstream.
+- Only One approval will be counted per each organization for a given PR.
+- The selection of which contributor to approve a PR is made by the WSL and should take those factors into consideration:
+  - Contributor needs to be actively discussing the PR to be selected for approval.
+  - Contributor needs to be active in the workstream.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: The previous rule states, "Contributor needs to be **actively discussing** the PR to be selected for approval." Does the fact that the contributor is **actively discussing** the issue not imply that they are "**active**" in the workstream? If not then we need to specify the requirements for being "active in the workstream."

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,65 +1,99 @@
...
 - PRs will be merged automatically online by the TSL once consensus is reached and all approvals are received. 
-- If WSL is the person who is creating the PR, They need to request an alternate approver, preferably a co-lead or from the committers list.
-
+- If WSL is the person who is creating the PR, They need to request an alternate approver, preferably a co-lead or from the contributors list.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: typo" delete "or" from "... a co-lead or from the ..."

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,65 +1,99 @@
...
 
 <p align="right"><img src="artefacts//figures/approval_process.png" alt="scope" title="Scope" width="100%"/></p>
 <p align="center"><b>Figure 1:</b> Approval Process</p>
 
+<a name="idle"></a>
+### Idle Pull Requests Policy
+
+Pull Requests will be automatically labled as "**Idle**" when:
+- No Engagement from the author of the PR to address Review Requests or input from the community for more than 15 calender days.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: what kind of "input" from the community? Can we simplify this and the following rule into simply, "No engagement/activity (content, reviews, conversations) from author, reviewers and workstream contributors on non-merge ready PRs for 15 calendar days."

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,65 +1,99 @@
...
+- One or more identified Reviewers (including WSL) are not providing feedback/resolutions or approving the Pull Request for more than 15 calender days.
+
+Any Pull Requests that are labeled as "Idle" will be discussed during Technical Steering Meeting to take a decision on, such as:
+- Close the PR and label it as "**Returned**".
+- "**Force**" Merging of the PR without having full reviewers approvals due to their inactivity.
+
+<a name="unresolved"></a>
+### Unresolved Pull Requests Policy
+
+Pull Requests will be automatically labled as "**Unresolved**" when no consensus are reached during approval process.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: typo: "... when no consensus are reached .." ---> "... when no consensus is reached  ..."

File: CONTRIBUTING.md (hunk 8 above)
Comment (hdamker) [reply CONTRIBUTING.md]: The label should be set "automatically" ... what does Github allow here?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -76,12 +78,19 @@ Pull Requests will be automatically labled as "**Idle**" when:
 - No Engagement from the author of the PR to address Review Requests or input from the community for more than 15 calender days.
   - Personal holidays or public holidays will not be counted.
```
Comment (hdamker) [inline CONTRIBUTING.md]: I guess that can't be judge automatically ... label should be set anyway, someone (e.g. WSL) can delete in such exceptions.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -76,12 +78,19 @@ Pull Requests will be automatically labled as "**Idle**" when:
...
 
 Any Pull Requests that are labeled as "Idle" will be discussed during Technical Steering Meeting to take a decision on, such as:
 - Close the PR and label it as "**Returned**".
 - "**Force**" Merging of the PR without having full reviewers approvals due to their inactivity.
-- Follow Governance Procedures to get consensus via agreement or voting mechanisms.
+
+<a name="unresolved"></a>
+### Unresolved Pull Requests Policy
+
+Pull Requests will be automatically labled as "**Unresolved**" when no consensus are reached during approval process.
```
Comment (hdamker) [inline CONTRIBUTING.md]: Can't be done automatically I suppose.

File: CONTRIBUTING.md (hunk 1 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: definition slightly modified.

Comment (rabi-abdel) [reply CONTRIBUTING.md]: I will be marking this as resolved.

File: CONTRIBUTING.md (hunk 2 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 3 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 4 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 5 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: reworded as suggested.

File: CONTRIBUTING.md (hunk 6 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: Good point, the second pint is deleted.

File: CONTRIBUTING.md (hunk 7 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: clarified. resolved

File: CONTRIBUTING.md (hunk 8 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: clarified automation for @hdamker and fixed type based on @pgoyal01 feedback

File: CONTRIBUTING.md (hunk 9 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: fixed.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -21,7 +22,8 @@ Following labels should be correctly used for all issues and PRs to indicate the
   - **WIP**: This label applies to PRs to indicate that this PR is still being developed and no reviews/approvals are still expected.
   - **Enhancement**: This label applies to both issues and PRs to indicate that the issue/Pr is proposing an Enhancement.
   - **Returned**: This label applies to both issues and PRs to indicate that the issue/PR has been returned for reasons such as (more information needed, not a high priority, no longer valid, etc).
-  - **Idle**: This label will be **automatically** applied to both issues/Labels to indicate that no activity has been observed on the issue/PR for some determined time frame. This an trigger an action to close the issue/PR.
+  - **Idle**: This label will be **automatically** applied to both issues/PRs to indicate that no activity has been observed on the issue/PR for some determined time frame. This an trigger an action to close the issue/PR.
+  - **Unresolved**: This label will be **automatically** applied to PRs to indicate that no consensus has been reached on the PR. This triggers an action to resolve the PR by the TSC.
```
[the remark this answers is not in the document]
Comment (rabi-abdel) [reply CONTRIBUTING.md]: @hdamker I clarified automatically. you are right, offourse the intention is to apply GitHub workflow which Is something we are dong right now. but until then, TSC will be manually applying those labels.

File: CONTRIBUTING.md (hunk 10 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: Yes I agree.

File: CONTRIBUTING.md (hunk 11 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: When I say automatically, I mean either find a way to do it in GitHub Workflow or TSC will be manually applying the label (automatically from the sense that they are not applied but the author or WSL).

**Comment by rabi-abdel** [comment]
@pgoyal01 @hdamker I updated the PR to reflect your review. please resolve your comments if you are happy with my changes and approve on the way :)

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,65 +1,98 @@
...
+  - **Chapter Number**: This label applies to PRs (and optionally to issues) to indicate which chapter the PR is targeting.
+  - **General**: This label applies to both issues and PRs to indicate that this issue/PR is not corresponding to any particular chapter/file.
+  - **Major/Minor**: This label applies to both issues and PRs to indicate if the issue/PR is addressing a major or minor change. 
+  - **WIP**: This label applies to PRs to indicate that this PR is still being developed and no reviews/approvals are expected.
+  - **Enhancement**: This label applies to both issues and PRs to indicate that the issue/Pr is proposing an Enhancement.
+  - **Returned**: This label applies to both issues and PRs to indicate that the issue/PR has been sent back to authoer (e.g. more information needed, not a high priority, no longer valid, etc).
+  - **Idle**: This label will be **automatically** applied (by a GitHub work flow or by TSC Leads) to both issues/PRs to indicate that no activity has been observed on the issue/PR for some determined time frame. This an trigger an action to close the issue/PR.
```
Comment (pgoyal01) [inline CONTRIBUTING.md]: Typo: "This an trigger an action to close the issue/PR." ---> "This triggers an action to close the issue/PR."

**Comment by pgoyal01** [comment]
@rabi-abdel Re: Figure 1. In the Step "Discuss and/or Vote" (between steps 5A and 6), I think one of the outcomes is possibly to return to Step 4  "PR being reviewed & edited" with guidance from the TSC (I think that is the reason for the "and/or" in "Discuss and/or Vote").

File: CONTRIBUTING.md (hunk 13 above)
Comment (rabi-abdel) [reply CONTRIBUTING.md]: fixed

**Comment by rabi-abdel** [comment]
> @rabi-abdel Re: Figure 1. In the Step "Discuss and/or Vote" (between steps 5A and 6), I think one of the outcomes is possibly to return to Step 4 "PR being reviewed & edited" with guidance from the TSC (I think that is the reason for the "and/or" in "Discuss and/or Vote").

if this is not major in your views, we can modify later since we are finalising the governance dolmens and can impact this diagram slightly to add more information about voting mechanisms. hope this is ok to defer editing the diagram until later.

**Pull request merged**